### PR TITLE
Demo submission: Automatically generate Python unittest results in HTML reports with Jenkins

### PR DESCRIPTION
**Pull request by Spycsh**
# Demo Submission: Automatically generate Python unittest results in HTML reports with Jenkins

## Members

Name: Sihan Chen,  Yuehao Sui

Mail: sihanc@kth.se, yuehao@kth.se

Github: Spycsh, amaothree

## Link

[Video](https://www.youtube.com/watch?v=emgC9u3xykk)

[GitHub repo related to the Video](https://github.com/Spycsh/python-unittest-htmlTestRunner-jenkins-demo)

[Another Take-home repo](https://github.com/Spycsh/python-unittest-htmlTestRunner-GitHub-Action-demo)

## Description
The demo will show how to automatically generate and visualize Python unittest results in HTML reports with Jenkins. The essential component to elegantly show the testing results in HTML reports is [HTMLTestRunner](https://github.com/SeldomQA/HTMLTestRunner), which is a non-trivial project itself and yield elegant HTML testing results for testers. The video will provide a simple example to show how it can be intergrated into Jenkins in a CI testing flow.

## Motivation
This is quite relevant to DevOps because jenkins is widely used for CI but there is no native support to generate HTML testing report for Python unittest. We hope this video demo will clearly show the whole flow.

## Criterias that we are aiming for

* The demonstration is clearly motivated
    - yes, as explained in the Motivation above

* The demonstration is difficult to do
    - yes & remarkable, because it is not easy to do from scratch and it depends on non-trivial infrastructure [HTMLTestRunner](https://github.com/SeldomQA/HTMLTestRunner) which is co-contributed by us. This infrastructure offers elegant HTML templates and also the mail support and the functionality to render testing results on the HTML. It is difficult to build it from scratch and there are no native support of such HTML template on Jenkins.

* The demonstration is original
    - yes & remarkable, although there are some videos on publishing HTML on jenkins, most of them are based on JUnit rather than python unittest. We also write our own [GitHub](https://github.com/Spycsh/python-unittest-htmlTestRunner-jenkins-demo) to show the whole flow and we promise that is original and non-duplicated.

* The video is sublime (eg. visually appealing)
    - yes, and HTML report is internally visually appealing than console output:)

* The video contains an easter egg
    - No

* There is a code repo to run the demo
    - yes & remarkable, see [GitHub](https://github.com/Spycsh/python-unittest-htmlTestRunner-jenkins-demo), the code repo is with a solid readme

* The video must contain subtitles which are clear and in proper English, and clear voice
    - yes & remarkable

* The video includes a take-home message
    - yes & remarkable, because you can deploy the whole flow on your own jenkins with the steps we wrote in README Of our [GitHub](https://github.com/Spycsh/python-unittest-htmlTestRunner-jenkins-demo), and there is also a take-home project you can try it out about how to use GitHub Action rather than jenkins to do the same things 
[Another Take-home repo](https://github.com/Spycsh/python-unittest-htmlTestRunner-GitHub-Action-demo).

### CODE REVIEW

**Comment by SophieHYe** [comment]
Thanks for the demo submission. I am now merging your PR.

**Comment by SophieHYe** [comment]
## Good points
* Overall very nice and clear demo.
* The voice over is very good and spoken clearly.
* The UI of the web app is good.
* Good to have subtitles. 
* Good to have some music for highlighting the content.
* Perfectly on time of 3 mins.

## To improve
* Give some motivation and explain to what extent it's related to DevOps.
* The font size of the code/text could be larger. Steps are hard to read and follow.
* Maybe give some reference reading materials to audience in this topic.

**Comment by andnil5** [comment]
**Overall, great job guys! I've added some comments and suggestions below.**
 
**Positive points**

* Nice that you highlighted things with boxes and zooming. It made it easier to understand where to look.
* In general, good and informative subtitles. It was easy to follow the demo and understand how to perform the actions.
* Nice introduction! I linked the popup with your contact details.
* Perfect in time!

**Possible improvements**

* It is nice that you have added a voice-over. However, the voice is very robotic, making it more of a distraction at some parts of the demo.
* The part about WSL2 and genie was a bit unclear. Also, what are the alternatives? It might be that I've not used Jenkins and it is straightforward if you have.
* At the end of the demo, you state:" For privacy reasons, I don't show it, it's simple". It may be reasonable not to show how to do it here, but it would have been helpful if there was a pointer to another source that explains how to do this then.

**Comment by evkade** [comment]
**Demo Feedback**

I first wrote a paragraph of general feedback and then divided more feedback into the same categories as the grading criteria and based it on which grading criteria you motivated you had fulfilled. 

**Overall feedback**

_Positive_ 

- The video is interesting and easy to follow.
- The instructions are good and clear
- The repo is really good 

_Negative_

- Sometimes, it feels like the video is going too fast.  
- You could maybe explain quickly about what Jenkins is and why it is used 
- Same with genie

**1. The demonstration is clearly motivated**

This criterion is fulfilled as the motivation is written in the PR, however, the motivation could also be added to the video or the code repo itself.

**2. The demonstration is difficult to do** 
**3. The demonstration is original**
 
Both are fulfilled and remarkable. Cool that you co-contributed to the tool that you showed in the video. 

**4. The video is sublime (eg. visually appealing)**
 
The problem here is that the video is a little bit blurry for me. 
 
**5. The video contains an easter egg**

Not applicable 

**6. There is a code repo to run the demo**
 
Fulfilled and remarkable: the code repo looks good, and the readme is very good and informative. 

**7. The video must contain subtitles which are clear and in proper English, and clear voice**
 
Fulfilled and remarkable. Maybe it is a little bit weird with the robot voice, but it still worked well. 

**8. The video includes a take-home message**
 
Fulfilled and remarkable: you motivated well why this criterion is fulfilled in your PR. I am thinking especially that the take-home project is a really good touch!

**Comment by vickstrom** [comment]
I was surprised how easy Jenkins was to setup! Will definitely make use of this in the future.

- Your video is visually nice. However, I felt that there was too much on the screen a few times. It would be nice if you guys could zoom in on the sections you're currently working at! 

- The subtitles are helpful, however, it covers a lot of the screen. I personally prefer youtube subtitles, but I guess it is personal preference. 

- The structure of the video is good. The e-mail report section at the end feels a bit out of context. Don't really understand why it was mentioned, because you decide not to show it. Maybe skip this and extend the analysis of the unit test results. 

Love the "nice" meme! Great video! :+1:

**Comment by sestys** [comment]
Nice demo! I found it relevant to the course and informative. I liked the use of the machine speaker, it is a good use if you are unsure of your pronunciation. Also, the repo is nicely structured. Few points to improve:

*  The mentioning of email felt unnecessary, even more when you decided not to show it because of privacy reasons - in that case, you should have used some generic information or blur personal stuff out.
* I missed more explanation why is this specifically relevant to DevOps in the video.

Overall, great job!

**Comment by nwessman** [comment]
# Feedback
Good video! It was visually very appealing. Adding random sound effects at relevant times made it fun to watch. It also explained the process very clearly.

This part is not explained, why change branch to main instead of master? (After rewatching it I notice that it is because your repo uses the main branch, I think this should be adressed.)

I would also prefer if you recorded your own voiceover instead of using a text-to-speech service, it made it less enjoyable to have to listen to a voice that is not very dynamic.

Overall a very good demo! It presented the information in a quick but clear way.  :trophy:

**Comment by hengque** [comment]
Overall I think you have made a good demo.
+ Clear walkthrough of what was done
+ Nice pacing, the demo never felt rushed
+ The text-to-speech ensured that the voiceover was clear at all times

- Would be good with background music to fill the silence of the pauses
- While your editing gives the video character, some of it made it harder to follow the video as a demo. More precisely, between 15 seconds and 30 seconds there is a lot of movement, zooming and changing pages which is a bit hard to follow.

**Comment by Spycsh** [comment]
@SophieHYe @andnil5 @evkade @vickstrom @sestys @nwessman @hengque Thank you for feedbacks! Many of feedbacks offer very valuable advices and we will surely consider about the improvement of our videos based on your suggestions and thank you again for watching and feedbacks!
